### ISSUE TITLE
File open leakage when opening an image from a stream (since v4.0.7)

### ISSUE THREAD

**Issue by pieleric** [wxWidgets/Phoenix#2025]
**Operating system**: Ubuntu 18.04 & 20.04 (other OS not tested)
**wxPython version & source**:   v4.0.7 and later, including 4.1.1 (pip & Ubuntu)
**Python version & source**:       v3.6 & v3.8 (Ubuntu)

**Description of the problem**:
The short story is that since v4.0.7, when opening an image like `wx.Image(stream)`, the stream is not closed afterwards, even if it's not used by anything else. Explicitly closing the stream does workaround the problem. See the example application for more specific details. It'd be great if we didn't have to explicitly close the stream.

The long story is that when upgrading to Ubuntu 20.04 our (large) GUI application started to break in all sort of unexpected ways. We've eventually found out that it's because after a while it runs out of opened file (the default Ubuntu limit is ~1024 files open per process). Further digging (with `lsof`) showed that it is because the GUI loads images via wx.Image(*stream*). Up to wxPython v4.0.6, this didn't happen.

Looking at the (relatively short) log, my first bet was that it is caused by commit f9c9004909, with the changes in etg/filesys.py, which look like this:
``` diff
--- a/etg/filesys.py
+++ b/etg/filesys.py
@@ -67,7 +67,9 @@ def run():
 
     c = module.find('wxFSFile')
     c.addPrivateCopyCtor()
-    c.find('wxFSFile.stream').transfer = True
-    c.find('DetachStream').transferBack = True
+    c.find('wxFSFile.stream').keepReference = True
```

Unfortunately, I've tried to reverse this change, and that didn't help. So either I mistook when doing the revert, or it's coming from somewhere else...

<details>
  <summary>Code Example (click to expand)</summary>

```python
#!/usr/bin/env python

import os
import wx
import psutil

IMAGE_FILE = os.path.dirname(wx.__file__) + "/py/PyCrust_32.png"

def ShowOpenedFiles():
    proc = psutil.Process()
    opened_files = proc.open_files()
    print(f"Now {len(opened_files)} files opened:\n{opened_files}")

def OpenImageFileName(fn):
    return wx.Image(fn)
    
def OpenImageStream(fn):
    # Leaking opened files, since wxPython 4.0.7
    f = open(fn, "rb")
    return wx.Image(f)

def JustOpenStream(fn):
    f = open(fn, "rb")

def OpenImageStreamClose(fn):
    f = open(fn, "rb")
    im = wx.Image(f)
    f.close()
    return im

app = wx.App()

print("Opening 5 images as a filename")
for i in range(5):
    im = OpenImageFileName(IMAGE_FILE)

del im
ShowOpenedFiles()

print("Opening 5 streams")
for i in range(5):
    JustOpenStream(IMAGE_FILE)

ShowOpenedFiles()

print("Opening 5 images via streams and close streams")
for i in range(5):
    im = OpenImageStreamClose(IMAGE_FILE)

del im
ShowOpenedFiles()

print("Opening 5 images via streams")
for i in range(5):
    im = OpenImageStream(IMAGE_FILE)

del im
ShowOpenedFiles()
```
It outputs the follow on 4.0.7 (and later):
```
Opening 5 images as a filename
Now 0 files opened:
[]
Opening 5 streams
Now 0 files opened:
[]
Opening 5 images via streams and close streams
Now 0 files opened:
[]
Opening 5 images via streams
Now 5 files opened:
[popenfile(path='/home/piel/alien/Phoenix/wx/py/PyCrust_32.png', fd=11, position=759, mode='r', flags=557056), popenfile(path='/home/piel/alien/Phoenix/wx/py/PyCrust_32.png', fd=12, position=759, mode='r', flags=557056), popenfile(path='/home/piel/alien/Phoenix/wx/py/PyCrust_32.png', fd=13, position=759, mode='r', flags=557056), popenfile(path='/home/piel/alien/Phoenix/wx/py/PyCrust_32.png', fd=14, position=759, mode='r', flags=557056), popenfile(path='/home/piel/alien/Phoenix/wx/py/PyCrust_32.png', fd=15, position=759, mode='r', flags=557056)]
```
On a earlier versions of wxPython, all functions calls are followed by `[]`.
</details>

**Comment by swt2c** [comment]
In my opinion, the 4.0.7+ behavior is correct.  If I pass a Python file-like object to <something>, I wouldn't expect <something> to close it.  BTW, I suspect the change in behavior was due to the change to `etg/stream.py` in f9c9004.

If you don't want to handle open/closing files, why don't you just use the `wx.Image(filename)` overload?

**Comment by pieleric** [comment]
I agree that wx.Image() shouldn't close automatically the file/stream. That's not what it used to do. It just didn't *hold a reference* to it after init. So the stream would get dereferenced once it's out of scope (ie, at the end of the function), and it would be automatically closed. Since v4.0.7, it seems that the wx.Image keeps a reference to the stream. Even if the wx.Image is dereferenced, the stream stays in memory, preventing it from ever automatically closing. The stream should only be needed during the wx.Image init, afterwards all the data has been loaded, and it's not needed anymore.

With `gc.get_referrers(f)`, you can see that there are references to the stream (even after calling `del im`).
```python
def OpenImageStreamClose(fn):
    f = open(fn, "rb")
    im = wx.Image(f)
    f.close()
    del im
    gc.collect()
    print(gc.get_referrers(f))
```
Prints:
```
[<frame object at 0x10ad7a8>, <built-in method read of _io.BufferedReader object at 0x7f7b36706b48>, <built-in method seek of _io.BufferedReader object at 0x7f7b36706b48>, <built-in method tell of _io.BufferedReader object at 0x7f7b36706b48>]
```

> why don't you just use the wx.Image(filename) overload?

Our application is actually doing things a little bit more complicated than this example. On some platform (Windows), we package the code and all the resources into a single executable, so the stream might might be some File-like object pointing to memory instead of a real File.

**Comment by pieleric** [comment]
> BTW, I suspect the change in behavior was due to the change to etg/stream.py in f9c9004.

Indeed! I've reverted this change too, and now it behaves as before.

**Comment by swt2c** [comment]
Okay, in that case, yes, that does sound wrong, but it is really a memory leak and not a file descriptor leak.

**Comment by pieleric** [comment]
Yes, you're very right. From our point of view, the annoying part was the file descriptor leakage induced by the memory leak. The workaround I've found fixes the file descriptor leakage, but indeed there is still a memory leak (which is not a big deal in our case, as a closed File takes very little memory). I'll update the bug title & description.
